```{r}
source('Rscripts/utils.R', chdir = TRUE)
```


```{r}
sample_tbl = "PS00227 PS00227-m6A-ML-val-gDNA-bc2001
PS00228 PS00228-m6A-ML-val-gDNA-40-U-EcoRI-bc2002
PS00229 PS00229-m6A-ML-val-gDNA-10-U-TaqI-bc2003
PS00230 PS00230-m6A-ML-val-gDNA-8-U-Dam-bc2004
PS00231 PS00231-m6A-ML-val-WGA-bc2005
PS00232 PS00232-m6A-ML-val-WGA-with-1-mM-m6ATP-bc2006
PS00234 PS00234-m6A-ML-val-WGA-with-160-uM-m6ATP-bc2008
PS00233 PS00233-m6A-ML-val-WGA-with-400-uM-m6ATP-bc2007
PS00235 PS00235-m6A-ML-val-WGA-with-64-uM-m6ATP-bc2009
PS00237 PS00237-m6A-ML-val-WGA-with-10-uM-m6ATP-bc2011
PS00236 PS00236-m6A-ML-val-WGA-with-25-uM-m6ATP-bc2010
PS00238 PS00238-m6A-ML-val-WGA-1U-Hia5-30-min-37-C-bc2012
PS00239 PS00239-m6A-ML-val-WGA-0.1U-Hia5-30-min-37-C-bc2013
PS00240 PS00240-m6A-ML-val-WGA-0.01U-Hia5-30-min-37-C-bc2014
PS00241 PS00241-m6A-ML-val-WGA-10-U-TaqI-bc2015
PS00242 PS00242-m6A-ML-val-WGA-8-U-Dam-bc2016
PS00243 PS00243-m6A-ML-val-Fiber-seq-bc2017
PS00244 PS00244-m6A-ML-val-Fiber-seq-M-SssI-bc2018
PS00245 PS00245-m6A-ML-val-WGA-M-SssI-bc2019"
sample_tbl = fread(sample_tbl, col.names=c("sm", "sample"))
```

```{r}
ms_tbl = "PS00231	-0.0305508
PS00232	67.944998
PS00233	52.2725014
PS00234	33.429858
PS00235	14.4525662
PS00236	8.4685369
PS00237	2.56382084"
ms_tbl = merge(fread(ms_tbl, col.names=c("sm", "m6a_percent")), sample_tbl, by="sm")
ms_tbl$type = "Mass spec"
ms_tbl
```


```{r}
ft_files = data.table(file=Sys.glob("/mmfs1/gscratch/stergachislab/bohaczuk/scripts/fiberseq-smk/results/P*subML/*tbl.gz"), type="fibertools")
ft_files = data.table(file=Sys.glob("/mmfs1/gscratch/stergachislab/bohaczuk/scripts/fiberseq-smk/results/P*_ML/*tbl.gz"), type="fibertools")
ipd_files = data.table(file=Sys.glob("/mmfs1/gscratch/stergachislab/bohaczuk/scripts/fiberseq-smk/results/P*subread/*tbl.gz"), type="ipdSummary")
files = bind_rows(ft_files, ipd_files) %>% 
    mutate(
       sm=str_remove(str_extract(file, "/PS\\d+"), "/")
    ) %>%
    merge(sample_tbl, by = "sm")
files
```

```{r}
get_m6a_percent = function(file) {
    tdf = fread(cmd = glue("hck -F sam_flag -F total_m6a_bp -F total_AT_bp -F ec -z {file}"))
    tdf  %>% 
        filter(sam_flag <= 16) %>%
        #filter(ec >= 8) %>%
        summarise(m6a=sum(total_m6a_bp), AT=sum(total_AT_bp), n_fibers = n()) %>% 
        mutate(
            m6a_percent = 100*m6a/AT
        )
}

m6a_per_df = files %>%
    filter(grepl("m6ATP|PS00231-m6A-ML-val-WGA-bc2005|Fiber-seq", sample)) %>%
    #head(2) %>%
    mutate(res = purrr::map(file, get_m6a_percent)) %>% 
    tidyr::unnest(res) %>%
    data.table()
```

```{r}
m6atp.ft = fread("validation_data/m6ATP.ft.tbl.gz")
new_data = m6atp.ft %>%
    group_by(RG) %>%
    summarise(
        m6a_percent = sum(total_m6a_bp)/sum(total_AT_bp) * 100,
        n_fibers = n(),
        m6a = sum(total_m6a_bp),
        AT = sum(total_AT_bp),
    ) %>%
    mutate(
        sm = str_remove(RG,"_ML.fiberseq"),
        type="fibertools",
    ) %>% 
    merge(sample_tbl, by="sm")
bind_rows(m6a_per_df, new_data)
```


```{r}
cur_colors = MODEL_COLORS[c("GMM","SEMI", "IPD")]
names(cur_colors) = c("ipdSummary", "fibertools", "Mass spec")
cur_labels = c(
    "Subread model",
    "Semi-supervised CNN",
    "Mass spectrometry"
)

ms_plot_df = bind_rows(m6a_per_df, ms_tbl, new_data) %>%
    filter(sm != "PS00243" & sm != "PS00244") %>%
    filter(grepl("with", sample)) %>%
    filter(!(type=="fibertools" & !is.na(file) )) %>%
    mutate(
        Hia5 = as.numeric(str_match(sample, ".*-([0-9]+)-(mM|uM)-m6ATP.*")[,2])
    ) %>%
    mutate(
        Hia5 = case_when(
            Hia5 == 1 ~ 1000,
            is.na(Hia5) ~ 1,
            TRUE ~ Hia5
        )
    ) %>%
    arrange(m6a_percent) %>%
    mutate(
        sm = factor(sm, levels=unique(sm))
    ) 

ms_ipd = ms_plot_df[type=="ipdSummary"]$m6a_percent
ms_ft = ms_plot_df[type=="fibertools"]$m6a_percent
ms_ms = ms_plot_df[type=="Mass spec"]$m6a_percent
cor.test(ms_ms, ms_ft)
cor.test(ms_ms, ms_ipd)

ms_plot_df %>%
    ggplot(aes(y = m6a_percent, x=Hia5, color=type)) +
    geom_line(size=.5) +
    geom_point(size=3, alpha=0.75) +
    geom_label_repel(
        aes(label = paste0(round(m6a_percent,1), "%")),
        min.segment.length=0,
        #alpha=0.75,
        #nudge_x = 1,
        direction="y",
        size=2.5,
        force=10,
        show.legend=FALSE,
        #force_pull=0.5,
    )+
    scale_x_continuous("WGA m6ATP concentration (uM)",
        trans="log10", #scales::pseudo_log_trans(),
        #breaks = c(1,10,100,1000),
        #labels = c("0", "10", "100", "1,000"),
        #minor_breaks = seq(1, 10, 1),
    ) +
    scale_y_continuous("Predicted percent m6A", trans="log10") +
    annotation_logticks(side="lb") +
    scale_color_manual("", values=cur_colors, labels=cur_labels) + 
    theme_minimal_grid() +
    theme(
        legend.position="top",
        legend.text=element_text(size=11),
    )
my_ggsave("Figures/m6a-percent.pdf", height=5, width=6)
```
# m6a percent version 2 with correlation
```{r}
library(ggpubr)
ms_plot_df %>%
    filter(type != "ipdSummary") %>%
    pivot_wider(
        id_cols = c("Hia5"),
        values_from = c(m6a_percent),
        names_from = c(type),
    ) %>%
    ggplot(aes(y = `Mass spec`, x=fibertools)) +
    geom_abline(aes(slope=1, intercept=0), linetype="dashed")+
    #geom_smooth(method = "lm", se=F) +
    geom_point(size=3, alpha=0.75) +
    geom_text(
        aes(x=4, y=5, label="y = x")
    )+
    stat_cor(r.digits = 3, p.digits = 3) +
    scale_x_continuous("Fibertools percent m6A",
       trans="log10", #scales::pseudo_log_trans(),
        #breaks = c(1,10,100,1000),
        #labels = c("0", "10", "100", "1,000"),
        #minor_breaks = seq(1, 10, 1),
    ) +
    scale_y_continuous("MS/MS percent m6A", trans="log10") +
    annotation_logticks(side="lb") +
    #scale_color_manual("", values=cur_colors, labels=cur_labels) + 
    theme_minimal_grid() +
    theme(
        #legend.position="top",
        legend.text=element_text(size=11),
    )
my_ggsave("Figures/m6a-percent-v2.pdf", height=5, width=6)
```

# ML calls vs threshold 
```{r}
libs = "export LIBTORCH=/mmfs1/gscratch/stergachislab/mvollger/projects/large_home/libs/libtorch-static-with-deps-1.13.0_cu116; export LIBTORCH_CXX11_ABI=0;export LD_LIBRARY_PATH=${LIBTORCH}/lib:$LD_LIBRARY_PATH;export DYLD_LIBRARY_PATH=${LIBTORCH}/lib:$LD_LIBRARY_PATH"

#ff = "validation_data/All_ML_PS00231-m6A-ML-val-WGA-bc2005.tbl.gz"
ff = "validation_data/tmp.tbl.gz"
WGA_data = "results/PS00092/unaligned.fiberseq.bam"
WGA_data = "validation_data/PS00231-m6A-ML-val-WGA-bc2005/PS00231_ML.fiberseq.bam"
cmd = glue("{libs} && samtools view -@ 16 -u -F 2304 -s 0.01 {WGA_data} | ft predict -t 20 -a -m 0 | ft extract -t 8 --all - | bgzip -@ 8 > {ff}")
print(cmd)
system(cmd)
```
```{r}
ff="validation_data/WGA.low.ml.ft.tbl.gz"
m6a_vs_ml_tbl = read_m6a(
    ff,
    ref=F,
    min_ml=200
) 
AT_bp = fread(glue("hck -z  -F sam_flag -F fiber -F total_AT_bp {ff}")) %>%
    filter(sam_flag <= 16)
dim(AT_bp)
```

```{r}
t=seq(min(m6a_vs_ml_tbl$m6a_qual, na.rm=T),255)
max_fpr = 20
m6a_vs_ml_tbl %>% 
    select(m6a_qual, start, en, st, fiber, tag, type) %>%
    head(n=1e6) %>%
    merge(AT_bp, by="fiber") %>%
    expand_grid(threshold=t)  %>%
    filter(m6a_qual >= threshold | is.na(m6a_qual)) %>%
    group_by(threshold) %>%
    summarise(
        m6a = sum(type=="m6A"),
        AT_bp = sum(unique(total_AT_bp))
    ) %>%
    mutate(
        m6a_frac = 100*m6a/AT_bp,
        ML_FPR = 100*(1.0 - threshold/256),  
    ) %>%
    filter(ML_FPR <= max_fpr) %>%
    #filter(type=="m6A") %>%
    ggplot(aes(y= m6a_frac, x=threshold/256*100)) +
    geom_line() +
    geom_point(
        #data = . %>% filter(threshold==target_ml_value),
    ) +
    geom_label_repel(
        data = . %>% filter(threshold==target_ml_value),
        aes(label=paste0(round(m6a_frac,2), "%")),
        min.segment.length=0,
        direction="x", 
        nudge_x=5,
        size=3
        #arrow=arrow,
    )+
    geom_vline(
        aes(xintercept=target_ml_value_per),
        linetype="dashed", 
        color="darkred",
    ) +
    #geom_abline(slope=1, intercept=0, color="darkred", linetype="dashed") +
    scale_x_continuous("Model estimated precision (%)") +
        #limits=c(0, max_fpr)
    #) +
    scale_y_continuous("Percent m6A in WGA data",
        #limits=c(0, max_fpr)
    ) + 
    #coord_fixed() +
    theme_minimal_grid()

my_ggsave("Figures/WGA-by-ML-tag.pdf", height=5, width=4)
```


# load npz files
```{r}
ipd_2.0 = load_ipd_from_npz("npz_r_data/2.0.npz")
ft_2.0 = load_ipd_from_npz("npz_r_data/ft_2.0.npz")
ipd_2.2 = load_ipd_from_npz("npz_r_data/2.2.npz")
ft_2.2 = load_ipd_from_npz("npz_r_data/ft_2.2.npz")
ipd_3.2 = load_ipd_from_npz("npz_r_data/3.2_fiberseq.npz")
ft_3.2 = load_ipd_from_npz("npz_r_data/ft_3.2.npz")
ft_Revio = load_ipd_from_npz("npz_r_data/Revio.npz")
ft_new_revio = load_ipd_from_npz("revio-training/t.bam.npz")
new_new_revio = load_ipd_from_npz("revio-training/tmp.revio.npz") 
#ipd_large_2.2 = load_ipd_from_npz("results/PS00109_2/ml/PS00109_2.npz")
#ipd_3.2_2 = load_ipd_from_npz("npz_r_data/3.2_AoU.npz")
#ipd_3.2_3 = load_ipd_from_npz("npz_r_data/3.2_m6ATP.npz")
long_ipd = bind_rows(
    list(
        ipd_2.0=ipd_2.0,
        ft_2.0=ft_2.0,
        ipd_2.2=ipd_2.2,
        ft_2.2=ft_2.2,
        ipd_3.2=ipd_3.2,
        ft_3.2=ft_3.2,
        #ipd_3.2_AoU=ipd_3.2_2,
        #ipd_3.2_m6ATP=ipd_3.2_3, # not a good one because dont have nucs to subset from
        #ipd_large_2.2=ipd_large_2.2,
        ft_Revio=ft_Revio,
        ft_new_revio=ft_new_revio,
        new_new_revio=new_new_revio
    ),
    .id = "chem"
)

filter_long_ipd = long_ipd %>%
    mutate(chem = factor(chem, levels=unique(chem))) %>%
    filter(name == "V8") %>%
    group_by(chem) %>%
    mutate(idx = seq(n())) %>%
    filter(idx < 1e6) 
dim(filter_long_ipd)
```

```{r}
filter_long_ipd %>%
    ggplot(aes(x=name, fill=label, y=255*value)) +
    geom_violin() +
    geom_text_repel(
        data = . %>% 
            group_by(chem, label, name) %>%
            summarise(
                my_mean = round(mean(255*value)),
                my_median = round(median(255*value)),
            ),
        aes(y=my_mean, label = my_mean),
        min.segment.length=0,
        nudge_y=15,
        nudge_x=-0.2,
    ) +
    facet_row(~chem) + 
    scale_y_continuous("IPD value") +
    xlab("PacBio chemistry") +
    scale_fill_discrete("ipdSummary m6A label") + 
    theme_minimal_grid() +  
    theme(
        legend.position="top"
    )
my_ggsave("Figures/ipd_values.pdf", width = 12)
```



```{r}
filter_long_ipd %>%
    group_by(chem, label, fiber, name) %>%
    summarise(
        value=mean(value),
        m6a_count = n()
    ) %>%
    #filter(m6a_count > 100) %>%
    ggplot(aes(x=name, fill=label, y=255*value)) +
    geom_violin(scale="width", width=0.75) +
    geom_label_repel(
        data = . %>% 
            group_by(chem, label, name) %>%
            summarise(
                my_mean = round(mean(255*value)),
                my_median = round(median(255*value)),
                fiber_count = length(unique(fiber)),
            ),
        aes(fill=NULL, y=my_mean, label = paste(my_median, "\nn =", comma(fiber_count))),
        alpha=0.75,
        min.segment.length=0,
        nudge_x=-2,
        nudge_y=20,
    ) +
    facet_row(~chem) + 
    scale_y_continuous("IPD value") +
    xlab("PacBio chemistry") +
    scale_fill_discrete("ipdSummary m6A label") + 
    theme_minimal_grid() +  
    theme(
        legend.position="top"
    )
my_ggsave("Figures/ipd_values_by_reads.pdf", width = 12)
```



```{r}
keep = filter_long_ipd %>%
    group_by(chem) %>%
    summarise(fiber = unique(fiber)) %>%
    mutate(index = seq(n())) %>%
    filter(index <= 200)

filter_long_ipd %>%
    filter(fiber %in% keep$fiber) %>%
    #filter(chem %in% c("ipd_2.0", "ipd_2.2", "ft_2.2", "ipd_3.2", "ft_Revio")) %>%
    group_by(chem, fiber, name) %>% 
    mutate(
        average=mean(value[label == TRUE]),
        m6a_count = sum(label==TRUE),
        m6a_frac = sum(label==TRUE)/n(),
    ) %>%
    ungroup() %>%
    #arrange(average) %>%
    arrange(m6a_count) %>%
    #arrange(m6a_frac) %>%
    mutate(
        fiber = factor(fiber, levels=unique(fiber)),
        #my_color = factor(label, levels=c(F,T)),
    ) %>%
    arrange(label) %>%
    ggplot(aes(x=fiber, fill=label, color=label, y=255*value)) +
    geom_point(aes(shape=label, alpha=(label+0.25) )) +
    geom_ysidedensity(position = "fill", adjust=0.5)+
    geom_xsidehistogram(position = "fill", stat="count", binwidth=1)+
    facet_col(~chem, scales = 'free') + 
    scale_y_continuous("IPD value") +
    scale_x_discrete("Fiber-seq reads", label=rep("",1000))+
    scale_alpha_continuous(limits=c(0,1))+
    xlab("PacBio chemistry") +
    scale_fill_discrete("ipdSummary m6A label") + 
    theme_minimal_grid() +  
    theme(
        legend.position="top"
    )
s=0.75
my_ggsave("Figures/ipd_values_per_read.pdf", width = 20*s, height=20*s)
```




```{r}
f = fread("validation_data/PS00243-m6A-ML-val-Fiber-seq-bc2017/PS00243_ML.fiberseq.all.tbl.gz")
i = fread("../fiberseq-smk/results/PS00243_subread/PS00243_subread.fiberseq.all.tbl.gz")
fi=bind_rows(list(f=f,i=i), .id="t")

fi %>% 
    filter(sam_flag <= 16) %>%
    group_by(t) %>%
    #filter(ec < 9) %>%
    summarise(
        100*sum(total_m6a_bp)/sum(total_AT_bp),
        100*median(total_m6a_bp/total_AT_bp),
        n()
    )
```



# kmer m6a counts 
```{r}
kmer.df = fread("Tables/m6a_kmers.tbl.gz")%>%
    mutate( 
        sm = gsub("_ML.fiberseq", "", RG),
        kmer_percent = size/raw_count * 100
    ) %>%
    merge(ms_tbl,by="sm", all.x=TRUE) %>%
    filter(size < raw_count) %>%
    filter(size > 0) %>%
    arrange(m6a_percent) %>%
    mutate(
        sm=factor(sm, levels=unique(sm))
    )
```

```{r}
p = kmer.df %>% 
    filter(sm == "PS00243") %>%
    ggplot( aes(fill=m6a_percent)) +
    geom_histogram(aes(x=kmer_percent),binwidth=0.5, color="black", alpha=0.75) +
    facet_col(~sm, scales="free_y")+
    scale_y_continuous("k-mer count", label=comma )+
    scale_x_continuous("Percent m6A", label=comma, limits=c(0,100))+
    scale_fill_continuous("MS/MS\n% m6A")+
    theme_minimal_grid()+
    theme(legend.position="right")
if(F){
    p = p + 
    geom_label_repel(
        data = kmer.df %>% group_by(sm, m6a_percent) %>% summarise,
        aes(x=m6a_percent, y=0, label=round(m6a_percent,1)), 
        size=2.5,
        alpha=0.85,
        fill="white",
        #color="darkred",
        nudge_y=10
    ) + 
    geom_vline(
        data = kmer.df %>% group_by(sm, m6a_percent) %>% summarise,
        aes(xintercept=m6a_percent), 
        linetype="dashed",
        color="red",
        size=0.75
    )
}
p
my_ggsave("Figures/kmer.pdf", height=8, width=12)
```